File: sys/teton/framebuffer/console.cpp
#include "framebuffer/console.h"

#include <mammoth/util/debug.h>

void Console::WriteChar(char c) {
  if (c == '\n') {
    CursorReturn();
    return;
  }

  uint64_t row = cursor_pos_ / cols();
  if (row >= rows()) {
    crash("Unimplemented console scroll.", glcr::UNIMPLEMENTED);
  }
  uint64_t fb_row = row * (psf_.height() + 1);
  uint64_t col = cursor_pos_ % cols();
  uint64_t fb_col = col * (psf_.width() + 1);

  uint8_t* glyph = psf_.glyph(c);

  for (uint32_t r = fb_row; r < fb_row + psf_.height(); r++) {
    for (uint32_t j = fb_col; j < fb_col + psf_.width(); j++) {
      uint8_t glyph_offset = psf_.width() - (j - fb_col) - 1;
      if ((glyph[r - fb_row] & (1 << glyph_offset))) {
        framebuf_.DrawPixel(r, j, 0xFFFFFFF);
      } else {
        framebuf_.DrawPixel(r, j, 0);
      }
    }
  }

  cursor_pos_++;
}

void Console::WriteString(glcr::StringView str) {
  for (uint64_t i = 0; i < str.size(); i++) {
    WriteChar(str[i]);
  }
}

void Console::CursorReturn() {
  cursor_pos_ -= cursor_pos_ % cols();
  cursor_pos_ += cols();
}

File: rust/sys/teton/src/console.rs
use crate::framebuffer::Framebuffer;
use crate::psf::Psf;

pub struct Console {
    framebuffer: Framebuffer,
    psf: Psf,
    row: u32,
    col: u32,
}

impl Console {
    pub fn new(framebuffer: Framebuffer, psf: Psf) -> Self {
        Self {
            framebuffer,
            psf,
            row: 0,
            col: 0,
        }
    }

    fn incr_cursor(&mut self) {
        self.col += 1;
        if self.col >= self.cols() {
            self.col = 0;
            self.row += 1;
        }

        if self.row >= self.rows() {
            panic!("Scroll unimplemented")
        }
    }

    pub fn write_char(&mut self, chr: char) {
        if chr == '\x08' {
            // Backspace.
            if self.col > 1 {
                self.col -= 1;
                self.write_char(' ');
                self.col -= 1;
            }
            return;
        }
        let glyph = self.psf.glyph(chr as u32);
        self.framebuffer.draw_glyph(
            glyph,
            self.row * (self.psf.height() + 1),
            self.col * (self.psf.width() + 1),
        );
        self.incr_cursor()
    }

    fn cols(&self) -> u32 {
        self.framebuffer.width() / (self.psf.width() + 1)
    }

    fn rows(&self) -> u32 {
        self.framebuffer.height() / (self.psf.height() + 1)
    }
}
